README.md
# ep_signalbar

Flutter widget to create signal bar, you can render signal-bar ui for your flutter project with this widget.

## Usage
To use this plugin, add `ep_signalbar` as a [dependency in your pubspec.yaml file](https://flutter.dev/platform-plugins/).

## Preview

![Image of Signalbar][signal]

[signal]: https://raw.githubusercontent.com/amhirin/ep_signalbar/master/assets/signal.gif

### Example

``` dart
import 'package:flutter/material.dart';
import 'package:ep_signalbar/ep_signalbar.dart';

void main() => runApp(MyApp());
class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Eduprog Signalbar Demo',
      home: MyHomePage(),
    );
  }
}
class MyHomePage extends StatefulWidget {
  MyHomePage({Key key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  EPSignalBar _sb1 = EPSignalBar(initBarValue: 5,);
  EPSignalBar _sb2 = EPSignalBar(
    barHeight: 100,
    barWidth: 200,
    barCount: 10,
    initBarValue: 8,
    animationCurve: Curves.easeOut,
    animationDuration: Duration(milliseconds: 200),
    barOnlineColor: Colors.blue,
    barOfflineColor: Colors.grey,
    barBorderColor: Colors.blue,
    barBackgroundColor: Colors.white.withOpacity(0.6),
  );
  EPSignalBar _sb3 = EPSignalBar(
    barHeight: 50,
    barWidth: 100,
    barCount: 4,
    initBarValue: 2,
    barOnlineColor: Colors.orangeAccent,
  );
  EPSignalBar _sb4 = EPSignalBar(
    barHeight: 50,
    barWidth: 100,
    barCount: 4,
  );
  EPSignalBar _sb5 = EPSignalBar(
    barHeight: 100,
    barWidth: 300,
    barCount: 15,
    initBarValue: 3,
    animationCurve: Curves.easeOut,
    animationDuration: Duration(milliseconds: 200),
    barOnlineColor: Colors.redAccent,
    barBackgroundColor: Colors.white.withOpacity(0.6),
  );

  EPSignalBar _sb6 = EPSignalBar(
    barHeight: 100,
    barWidth: 300,
    barCount: 10,
    initBarValue: 7,
    animationDuration: Duration(milliseconds: 200),
    barOnlineColor: Colors.green,
    barBackgroundColor: Colors.grey.withOpacity(0.4),
    barBorderColor: Colors.green,
  );

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            FlatButton(
              child: Text("Set Bar Change"),
              color: Colors.red,
              onPressed: (){
                if (_sb1.state.getBarNumber() > 0){
                  _sb1.state.setBarNumber(_sb1.state.getBarNumber() - 1);
                }else{
                  _sb1.state.setBarNumber(5);
                }

              },
            ),
            _sb1,
            _sb2,
            _sb3,
            _sb4,
            _sb5,
            _sb6
          ],
        ),
      ), 
    );
  }
}

```


### Cheers

A.M Hirin

(eduprog.net)

CHANGELOG.md
## [0.0.1] - 2020-01-24.

First release of component

## [0.0.2] - 2020-01-24.

Update readme

## [0.0.3] - 2020-01-29.

Update example


lib/ep_signalbar.dart
library ep_signalbar;

import 'dart:ui';
import 'package:flutter/material.dart';

class EPSignalBar extends StatefulWidget {
  Color  barOnlineColor;
  Color  barOfflineColor;
  Color  barBackgroundColor;
  Color  barBorderColor;
  int barCount = 0;
  int initBarValue = 0;
  double barHeight = 0;
  double barWidth = 0;
  Duration animationDuration;
  Curve animationCurve;

  _EPSignalBarState state;

  EPSignalBar({Key key,
    this.barOnlineColor, this.barOfflineColor,
    this.barBackgroundColor, this.barBorderColor,
    this.barCount, this.barHeight,
    this.barWidth, this.animationDuration,
    this.animationCurve, this.initBarValue


  }) : super(key: key);

  @override
  _EPSignalBarState createState(){
    return state = _EPSignalBarState();
  }
}
class _EPSignalBarState extends State<EPSignalBar> {

  int _curBar = 0;

  double _getBarHeight(int barIndex){
    return _getHPercent(15 * (5 / widget.barCount)) + (barIndex * _getHPercent(10 * (5 / widget.barCount)));
  }

  double _getBarWidth(){
    return ((widget.barWidth - 20) / widget.barCount) - _getWPercent(6 * (5 / widget.barCount));
  }

  double _getBarCircular(){
    return  _getBarWidth() * (22/100);
  }

  Widget buildBarWidget(){
    List<AnimatedContainer> lstBar = [];
    for (int i = 0; i < widget.barCount; i++){
      AnimatedContainer _a = AnimatedContainer(
          duration: widget.animationDuration,
          curve: widget.animationCurve,
          margin: (widget.barCount - i ) > 1 ? EdgeInsets.only(right: _getWPercent(4 * (5 / widget.barCount))) : EdgeInsets.only(right:0),
          height: _getBarHeight(i),
          width: _getBarWidth(),
          decoration: BoxDecoration(
            color:  (_curBar - 1)  < i ? widget.barOfflineColor : widget.barOnlineColor,
            borderRadius: BorderRadius.only(
                topLeft: Radius.circular(_getBarCircular()),
                topRight: Radius.circular(_getBarCircular())
            ),
          )
      );
      lstBar.add(_a);
    }
    return Container(
      //alignment: Alignment.center,
      padding: EdgeInsets.all(10),
      height: widget.barHeight,
      width: widget.barWidth,
      decoration: BoxDecoration(
          color: widget.barBackgroundColor,
          border: Border.all(width: 2, color: widget.barBorderColor),
          borderRadius: BorderRadius.only(

          )
      ),
      child: Row(
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.end,
        children: lstBar,
      ),
    );
  }
  @override

  Widget build(BuildContext context) {
    return buildBarWidget();
  }

  void setBarNumber(int no){
    setState(() {
      if (no < 0) no = 0;
      if (no > widget.barCount) no = widget.barCount;
      _curBar = no;
    });
  }

  int getBarNumber(){
    return _curBar;
  }

  void setWidgetHeight(double v){
    if (v < 40) v = 40; //. min height
    setState(() {
      widget.barHeight = v;
    });
  }

  void setWidgetWidth(double v){
    if (v < 40) v = 40; //. min width
    setState(() {
      widget.barWidth = v;
    });
  }

  double _getHPercent(double p){
    return (p / 100) * widget.barHeight;
  }

  double _getWPercent(double p){
    return (p / 100) * widget.barWidth;
  }

  @override
  void initState() {
    //. set default value if null
    if (widget.barCount == null) widget.barCount = 0;
    if (widget.barHeight == null) widget.barHeight = 0;
    if (widget.barWidth == null) widget.barWidth = 0;
    if (widget.initBarValue == null) widget.initBarValue = 0;
    if (widget.animationCurve == null) widget.animationCurve = Curves.easeIn;
    widget.barOnlineColor = widget.barOnlineColor ?? Color.fromRGBO(0, 140, 0, 1.0);
    widget.barOfflineColor = widget.barOfflineColor ?? Color.fromRGBO(125, 125, 125, 1.0);
    widget.barBackgroundColor = widget.barBackgroundColor ?? Colors.transparent;
    widget.barBorderColor = widget.barBorderColor ?? Colors.transparent;
    widget.barCount = widget.barCount > 0 ? widget.barCount : 5 ;
    widget.animationDuration = widget.animationDuration ?? Duration(milliseconds: 300);
    widget.barHeight = widget.barHeight >= 40 ? widget.barHeight : 100 ;
    widget.barWidth = widget.barWidth >= 40 ? widget.barWidth : 130;
    widget.animationCurve = widget.animationCurve  ;
    if (widget.initBarValue >= 0 && widget.initBarValue <= widget.barCount){
      _curBar = widget.initBarValue;
    }else{
      _curBar = widget.barCount;
    }


    super.initState();
  }
}


example/lib/main.dart
import 'dart:async';
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:ep_signalbar/ep_signalbar.dart';


void main() => runApp(MyApp());
class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Eduprog Signalbar Demo',
      home: MyHomePage(),
      debugShowCheckedModeBanner: false,
    );
  }
}
class MyHomePage extends StatefulWidget {
  MyHomePage({Key key}) : super(key: key);

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  EPSignalBar _sb1 = EPSignalBar(initBarValue: 5,);
  EPSignalBar _sb2 = EPSignalBar(
    barHeight: 100,
    barWidth: 200,
    barCount: 10,
    initBarValue: 8,
    animationCurve: Curves.easeOut,
    animationDuration: Duration(milliseconds: 200),
    barOnlineColor: Colors.blue,
    barOfflineColor: Colors.grey,
    barBorderColor: Colors.blue,
    barBackgroundColor: Colors.white.withOpacity(0.3),
  );
  EPSignalBar _sb3 = EPSignalBar(
    barHeight: 50,
    barWidth: 100,
    barCount: 4,
    initBarValue: 2,
    barOnlineColor: Colors.orangeAccent,
  );
  EPSignalBar _sb4 = EPSignalBar(
    barHeight: 50,
    barWidth: 100,
    barCount: 4,
    barOnlineColor: Colors.purple,
    barOfflineColor: Colors.black,
  );
  EPSignalBar _sb5 = EPSignalBar(
    barHeight: 100,
    barWidth: 300,
    barCount: 15,
    initBarValue: 3,
    animationCurve: Curves.easeOut,
    animationDuration: Duration(milliseconds: 200),
    barOnlineColor: Colors.redAccent,
    barBackgroundColor: Colors.white.withOpacity(0.6),
  );

  EPSignalBar _sb6 = EPSignalBar(
    barHeight: 100,
    barWidth: 300,
    barCount: 10,
    initBarValue: 7,
    animationDuration: Duration(milliseconds: 200),
    barOnlineColor: Colors.green,
    barBackgroundColor: Colors.grey.withOpacity(0.4),
    barBorderColor: Colors.green,
  );

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: SingleChildScrollView(
          child: Column(
            children: <Widget>[
              Center(
                child: Column(

                  mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                  mainAxisSize: MainAxisSize.max,

                  children: <Widget>[
                    Container(
                      padding: EdgeInsets.only(top: 20),
                      child: Text("Eduprog Widget Signalbar", style: TextStyle(color: Colors.red, fontSize: 20, fontWeight: FontWeight.bold),),
                    ),
                    _sb1,
                    _sb2,
                    _sb3,
                    _sb4,
                    _sb6
                  ],
                ),
              ),
              Container(
                height: 300,
                child: _sb5,

              )
            ],
          ),
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }

  @override
  void initState() {
    Timer.periodic(Duration(milliseconds: 200), (timer) {
      if (_sb2.state.getBarNumber() > 1){
        _sb2.state.setBarNumber(_sb2.state.getBarNumber() - 1);
      }else{
        _sb2.state.setBarNumber(_sb2.barCount);
      }

      if (_sb5.state.getBarNumber() > 1){
        _sb5.state.setBarNumber(_sb5.state.getBarNumber() - 1);
      }else{
        _sb5.state.setBarNumber(_sb5.barCount);
      }

    });
    Timer.periodic(Duration(milliseconds: 1000), (timer) {
      if (_sb3.state.getBarNumber() == 3){
        _sb3.state.setBarNumber(2);
      }else{
        _sb3.state.setBarNumber(3);
      }

      if (_sb4.state.getBarNumber() > 1){
        _sb4.state.setBarNumber(_sb4.state.getBarNumber() - 1);
      }else{
        _sb4.state.setBarNumber(_sb4.barCount);
      }

      if (_sb6.state.getBarNumber() == 2){
        _sb6.state.setBarNumber(4);
      }else if (_sb6.state.getBarNumber() == 4){
        _sb6.state.setBarNumber(10);
      }else if (_sb6.state.getBarNumber() == 10){
        _sb6.state.setBarNumber(5);
      }else if (_sb6.state.getBarNumber() == 5){
        _sb6.state.setBarNumber(0);
      }else{
        _sb6.state.setBarNumber(2);
      }

      if (_sb5.barHeight > 200){
        _sb5.state.setWidgetHeight(200);
      }else{
        _sb5.state.setWidgetHeight(250);
      }
    });

    super.initState();
  }

  @override
  void setState(fn) {
    if(mounted){
      super.setState(fn);
    }
  }
}

